ctk/common/src/test/kotlin/org/codice/compilance/verification/core/CoreVerifierSpec.kt

```diff
@@ -0,0 +1,122 @@
+        every { specRefMsg.name } returns "TEST"
+
+        "single error status response has expected code" {
+            val statusCode = "urn:samlconf:fail"
+
+            buildDom(response(statusCode)).let {
+                CoreVerifier.verifyErrorStatusCode(it, specRefMsg, statusCode)
+            }
+        }
+
+        "mismatched status code fails and does not contain SAMLCore 3.2.1 response" {
+            buildDom(response("urn:samlconf:fail")).let {
+                shouldThrow<SAMLComplianceException> {
+                    CoreVerifier.verifyErrorStatusCode(it, specRefMsg, "urn:samlconf:expected")
+                }.message?.shouldNotContain(SAMLCore_3_2_1_d.message)
+            }
+        }
+
+        "failure with expected code of REQUESTER includes SAMLCore 3.2.1 response" {
+            buildDom(response("urn:samlconf:fail")).let {
+                shouldThrow<SAMLComplianceException> {
+                    CoreVerifier.verifyErrorStatusCode(it, specRefMsg, TestCommon.REQUESTER)
+                }.message?.shouldContain(SAMLCore_3_2_1_d.message)
+            }
+        }
+
+        "validateTimeWindow with null datetime attributes" {
+            buildDom("<fld>this is garbage</fld>").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+
+            buildDom("""<fld NotBefore="$now"/>""").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+
+            buildDom("""<fld NotOnOrAfter="$now"/>""").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+
+            buildDom("""<fld NotBefore="$later"/>""").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+
+            buildDom("""<fld NotOnOrAfter="$later"/>""").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+        }
+
+        "validateTimeWindow with correct datetime attributes" {
+            buildDom("""<fld NotBefore="$now" NotOnOrAfter="$later"/>""").let {
+                CoreVerifier.validateTimeWindow(it, specRefMsg)
+            }
+        }
+
+        "validateTimeWindow with incorrect datetime attributes" {
+            buildDom("""<fld NotBefore="$later" NotOnOrAfter="$now"/>""").let {
+                shouldThrow<SAMLComplianceException> {
+                    CoreVerifier.validateTimeWindow(it, specRefMsg)
+                }
+            }
+
+            buildDom("""<fld NotBefore="$now" NotOnOrAfter="$now"/>""").let {
+                shouldThrow<SAMLComplianceException> {
+                    CoreVerifier.validateTimeWindow(it, specRefMsg)
+                }
+            }
+
+            buildDom("""<fld NotBefore="$later" NotOnOrAfter="$later"/>""").let {
+                shouldThrow<SAMLComplianceException> {
+                    CoreVerifier.validateTimeWindow(it, specRefMsg)
+                }
+            }
+        }
+    }
```

Comment:
❗️ Technically you might want to include the negative path of having only one of `NotBefore` or `NotOnOrAfter` and the value of it being invalid. You could assert that the SamlComplianceException shouldNotContain the specRefMsg since `CommonDataTypeVerifier.verifyDateTimeValues(this)` will not include the passed in message.

Reply:
I just had a realization from reading this comment that I wanted to share (correct me if I'm wrong 😨):

It would be overkill / bad practice to add specific exception testing for the delegated methods. For example: making sure that when we provide a bad data type, we get `SAMLCore_1_3_3_a` in the exception message.

However, if we know that those methods should always throw an exception, catching a general exception would be ideal for testing that the method will error out. That way we can change the other class's exception as much as necessary without needing to changing this seemingly unrelated test class.

Reply:
Yeah it would not be a very good practice, that's true. Ideally we'd be able to verify that the method we're testing ends up calling `CommonDataTypeVerifier.verifyDateTimeValues(this)`, but it being a static method call makes that difficult. I guess expecting any SamlComplianceException would be fine.

Reply:
I'm not sure what you mean by having only one and it not being valid. Having only one gets tested in several permutations in `validateTimeWindow with null datetime attributes` and behaves as expected.